common/web/lm-worker/src/main/model-compositor.ts

```diff
@@ -181,24 +130,35 @@ export default class ModelCompositor {
 
     // We want to dedupe before trimming the list so that we can present a full set
     // of viable distinct suggestions if available.
-    let deduplicatedSuggestionTuples = this.dedupeSuggestions(rawPredictions, context);
+    let deduplicatedSuggestionTuples = dedupeSuggestions(this.lexicalModel, rawPredictions, context);
 
     // Needs "casing" to be applied first.
     //
     // Will also add a 'keep' suggestion (with `.matchesModel = false`) matching
     // the current state of context if there is no such matching prediction.
-    this.processSimilarity(deduplicatedSuggestionTuples, context, transformDistribution[0]);
+    processSimilarity(this.lexicalModel, deduplicatedSuggestionTuples, context, transformDistribution[0]);
 
     // Section 3:  Sort the suggestions in display priority order to determine
     // which are most optimal, then auto-select based on the results.
     deduplicatedSuggestionTuples = deduplicatedSuggestionTuples.sort(tupleDisplayOrderSort);
-    this.predictionAutoSelect(deduplicatedSuggestionTuples);
+    predictionAutoSelect(deduplicatedSuggestionTuples);
 
     // Section 4: Trim down the suggestion list to the N (MAX_SUGGESTIONS) most optimal,
     // then cache and return the set of suggestions
 
     // Now that we've marked the suggestion to auto-select, we can finalize the suggestions.
-    const suggestions = this.finalizeSuggestions(deduplicatedSuggestionTuples, context, inputTransform);
+    const suggestions = finalizeSuggestions(
+      this.lexicalModel,
+      deduplicatedSuggestionTuples.splice(0, ModelCompositor.MAX_SUGGESTIONS),
+      context,
+      inputTransform,
+      this.verbose
+    );
+
+    suggestions.forEach((suggestion) => {
+      suggestion.id = this.SUGGESTION_ID_SEED;
+      this.SUGGESTION_ID_SEED++;
+    });
```

Comment:
The suggestion unique-identifier tagging process was moved out of `finalizeSuggestions` for similar reasons:  generating unique IDs is inherently stateful.